File: src/profile/annotate.ts
import type { ServiceIR, OperationDecl, ParamDecl } from "#ir/types.js";
import type {
  ProfiledService,
  ProfiledEntity,
  ProfiledOperation,
  ProfiledField,
  DeploymentProfile,
} from "#profile/types.js";
import { getProfile } from "#profile/registry.js";
import { mapType, type TypeContext } from "#profile/type-map.js";
import { classifyOperations } from "#convention/classify.js";
import { deriveEndpoints } from "#convention/path.js";
import { deriveSchema } from "#convention/schema.js";
import { toSnakeCase, toTableName, toColumnName, pluralize } from "#convention/naming.js";
import { getConvention } from "#convention/path.js";

export function buildProfiledService(ir: ServiceIR, profileName: string): ProfiledService {
  const profile = getProfile(profileName);
  const classifications = classifyOperations(ir);
  const endpoints = deriveEndpoints(classifications, ir);
  const schema = deriveSchema(ir);

  const ctx: TypeContext = {
    entityNames: new Set(ir.entities.map((e) => e.name)),
    enumNames: new Set(ir.enums.map((e) => e.name)),
    aliasMap: new Map(ir.typeAliases.map((a) => [a.name, a.typeExpr])),
  };

  const classificationMap = new Map(classifications.map((c) => [c.operationName, c]));
  const endpointMap = new Map(endpoints.map((e) => [e.operationName, e]));
  const tableMap = new Map(schema.tables.map((t) => [t.entityName, t]));

  const entities = ir.entities.map((entity) => {
    const table = tableMap.get(entity.name);
    const tableNameOverride = getConvention(ir.conventions, entity.name, "db_table");
    const tableName = tableNameOverride || toTableName(entity.name);

    return profileEntity(entity.name, tableName, entity.fields, profile, ctx, table !== undefined);
  });

  const operations = ir.operations.map((op) => {
    const classification = classificationMap.get(op.name);
    const endpoint = endpointMap.get(op.name);
    if (!classification || !endpoint) {
      return {
        operationName: op.name,
        handlerName: toSnakeCase(op.name),
        endpoint: endpoint!,
        kind: classification?.kind ?? ("side_effect" as const),
        targetEntity: classification?.targetEntity ?? null,
        requestBodyFields: profileParams(op.inputs, profile, ctx),
        responseFields: profileParams(op.outputs, profile, ctx),
      };
    }

    return profileOperation(op, classification.kind, classification.targetEntity, endpoint, profile, ctx);
  });

  return { ir, profile, endpoints, schema, entities, operations };
}

function profileEntity(
  entityName: string,
  tableName: string,
  fields: readonly { readonly name: string; readonly typeExpr: import("#ir/types.js").TypeExpr }[],
  profile: DeploymentProfile,
  ctx: TypeContext,
  _hasTable: boolean,
): ProfiledEntity {
  const snakeName = toSnakeCase(entityName);
  const pluralSnake = toSnakeCase(pluralize(entityName));

  const profiledFields = fields.map((f) => profileField(f.name, f.typeExpr, profile, ctx));

  return {
    entityName,
    tableName,
    modelClassName: entityName,
    createSchemaName: `${entityName}Create`,
    readSchemaName: `${entityName}Read`,
    updateSchemaName: `${entityName}Update`,
    modelFileName: `${snakeName}.py`,
    schemaFileName: `${snakeName}.py`,
    routerFileName: `${pluralSnake}.py`,
    fields: profiledFields,
  };
}

function profileField(
  fieldName: string,
  typeExpr: import("#ir/types.js").TypeExpr,
  profile: DeploymentProfile,
  ctx: TypeContext,
): ProfiledField {
  const mapped = mapType(typeExpr, profile, ctx);
  const colName = toColumnName(fieldName);
  const nullable = typeExpr.kind === "OptionType";
  const sqlalchemyColumnType = resolveColumnType(typeExpr, profile, ctx);

  return {
    fieldName,
    columnName: colName,
    pythonType: mapped.python,
    pydanticType: mapped.pydantic,
    sqlalchemyType: mapped.sqlalchemy,
    sqlalchemyColumnType,
    nullable,
    hasDefault: false,
  };
}

function resolveColumnType(
  typeExpr: import("#ir/types.js").TypeExpr,
  profile: DeploymentProfile,
  ctx: TypeContext,
): string {
  switch (typeExpr.kind) {
    case "NamedType": {
      const mapping = profile.typeMap.get(typeExpr.name);
      if (mapping) return mapping.sqlalchemy;
      if (ctx.entityNames.has(typeExpr.name)) return "Integer";
      if (ctx.enumNames.has(typeExpr.name)) return "String";
      const alias = ctx.aliasMap.get(typeExpr.name);
      if (alias) return resolveColumnType(alias, profile, ctx);
      return "String";
    }
    case "OptionType":
      return resolveColumnType(typeExpr.innerType, profile, ctx);
    case "SetType":
    case "SeqType":
    case "MapType":
      return "JSONB";
    case "RelationType":
      return "Integer";
  }
}

function profileOperation(
  op: OperationDecl,
  kind: import("#convention/types.js").OperationKind,
  targetEntity: string | null,
  endpoint: import("#convention/types.js").EndpointSpec,
  profile: DeploymentProfile,
  ctx: TypeContext,
): ProfiledOperation {
  return {
    operationName: op.name,
    handlerName: toSnakeCase(op.name),
    endpoint,
    kind,
    targetEntity,
    requestBodyFields: profileParams(op.inputs, profile, ctx),
    responseFields: profileParams(op.outputs, profile, ctx),
  };
}

function profileParams(
  params: readonly ParamDecl[],
  profile: DeploymentProfile,
  ctx: TypeContext,
): readonly ProfiledField[] {
  return params.map((p) => profileField(p.name, p.typeExpr, profile, ctx));
}

File: modules/profile/src/main/scala/specrest/profile/Annotate.scala
package specrest.profile

import specrest.convention.Classify
import specrest.convention.EndpointSpec
import specrest.convention.Path
import specrest.convention.Schema
import specrest.ir.*
import specrest.ir.generated.SpecRestGenerated.*

@SuppressWarnings(Array("org.wartremover.warts.IsInstanceOf"))
object Annotate:

  def buildProfiledService(ir: ServiceIRFull, profileName: String): ProfiledService =
    val profile         = Registry.getProfile(profileName)
    val classifications = Classify.classifyOperations(ir)
    val endpoints       = Path.deriveEndpoints(classifications, ir)
    val schema          = Schema.deriveSchema(ir)

    val ix = ir.idx
    val ctx = TypeContext(
      entityNames = ix.entityNames,
      enumNames = ix.enumNames,
      aliasMap = ix.aliases.map(a => talName(a) -> talType(a)).toMap
    )

    val classificationMap = classifications.map(c => classificationOperationName(c) -> c).toMap
    val endpointMap       = endpoints.map(e => e.operationName -> e).toMap
    val tableMap          = schemaTables(schema).map(t => tableEntityName(t) -> t).toMap

    val entities = ix.entities.map { entity =>
      val tableName = Path
        .getConvention(svcConventions(ir), entName(entity), "db_table")
        .getOrElse(Naming.toTableName(entName(entity)))
      profileEntity(
        entName(entity),
        tableName,
        entFields(entity),
        profile,
        ctx,
        tableMap.contains(entName(entity))
      )
    }

    val operations = svcOperations(ir).map { op =>
      val classification = classificationMap(operName(op))
      val endpoint       = endpointMap(operName(op))
      profileOperation(
        op,
        classificationKind(classification),
        classificationTargetEntity(classification),
        endpoint,
        profile,
        ctx
      )
    }

    ProfiledService(ir, profile, endpoints, schema, entities, operations)

  def attachDafnyMethods(
      profiled: ProfiledService,
      bindings: Map[String, String]
  ): ProfiledService =
    if bindings.isEmpty then profiled
    else
      profiled.copy(operations =
        profiled.operations.map: op =>
          bindings.get(op.operationName) match
            case Some(callable) => op.copy(dafnyMethod = Some(callable))
            case None           => op
      )

  private def profileEntity(
      entityName: String,
      tableName: String,
      fields: List[field_decl],
      profile: DeploymentProfile,
      ctx: TypeContext,
      hasTable: Boolean
  ): ProfiledEntity =
    val _            = hasTable
    val snakeName    = Naming.toSnakeCase(entityName)
    val pluralSnake  = Naming.toSnakeCase(Naming.pluralize(entityName))
    val profiledFlds = fields.map(f => profileField(fldName(f), fldType(f), profile, ctx))
    ProfiledEntity(
      entityName = entityName,
      tableName = tableName,
      modelClassName = entityName,
      createSchemaName = s"${entityName}Create",
      readSchemaName = s"${entityName}Read",
      updateSchemaName = s"${entityName}Update",
      modelFileName = s"$snakeName.py",
      schemaFileName = s"$snakeName.py",
      routerFileName = s"$pluralSnake.py",
      fields = profiledFlds
    )

  private def profileField(
      fieldName: String,
      typeExpr: type_expr,
      profile: DeploymentProfile,
      ctx: TypeContext
  ): ProfiledField =
    val mapped   = TypeMap.mapType(typeExpr, profile, ctx)
    val colName  = Naming.toColumnName(fieldName)
    val resolved = TypeMap.resolveTypeExpr(typeExpr, ctx.aliasMap)
    val nullable = resolved.isInstanceOf[OptionTypeF]
    val columnType =
      widenExplicitIdPkSqlType(fieldName, resolveColumnType(typeExpr, profile, ctx))
    ProfiledField(
      fieldName = fieldName,
      columnName = colName,
      domainType = mapped.domain,
      validationType = mapped.validation,
      ormFieldType = mapped.orm,
      ormColumnType = columnType,
      nullable = nullable,
      hasDefault = false
    )

  private def resolveColumnType(
      typeExpr: type_expr,
      profile: DeploymentProfile,
      ctx: TypeContext
  ): String =
    val defaults = ColumnTypeDefaults.forProfile(profile)
    typeExpr match
      case NamedTypeF(name, _) =>
        profile.typeMap.get(name) match
          case Some(m)                                => m.ormColumn
          case None if ctx.entityNames.contains(name) => defaults.relation
          case None if ctx.enumNames.contains(name)   => defaults.enum_
          case None =>
            ctx.aliasMap.get(name) match
              case Some(alias) => resolveColumnType(alias, profile, ctx)
              case None        => defaults.fallback
      case OptionTypeF(inner, _)                               => resolveColumnType(inner, profile, ctx)
      case SetTypeF(_, _) | SeqTypeF(_, _) | MapTypeF(_, _, _) => defaults.collection
      case RelationTypeF(_, _, _, _)                           => defaults.relation

  private case class ColumnTypeDefaults(
      relation: String,
      enum_ : String,
      fallback: String,
      collection: String
  )

  private object ColumnTypeDefaults:
    private val Python     = ColumnTypeDefaults("Integer", "String", "String", "JSONB")
    private val Go         = ColumnTypeDefaults("BIGINT", "TEXT", "TEXT", "JSONB")
    private val TypeScript = ColumnTypeDefaults("INTEGER", "TEXT", "TEXT", "JSONB")

    def forProfile(profile: DeploymentProfile): ColumnTypeDefaults =
      profile.language match
        case "go" => Go
        case "ts" => TypeScript
        case _    => Python

  private def profileOperation(
      op: operation_decl,
      kind: operation_kind,
      targetEntity: Option[String],
      endpoint: EndpointSpec,
      profile: DeploymentProfile,
      ctx: TypeContext
  ): ProfiledOperation =
    ProfiledOperation(
      operationName = operName(op),
      handlerName = Naming.toSnakeCase(operName(op)),
      endpoint = endpoint,
      kind = kind,
      targetEntity = targetEntity,
      requestBodyFields =
        operInputs(op).map(p => profileField(prmName(p), prmType(p), profile, ctx)),
      responseFields =
        operOutputs(op).map(p => profileField(prmName(p), prmType(p), profile, ctx)),
      requiresAuth = operRequiresAuth(op).getOrElse(Nil)
    )
